packages/delegation-process/src/services/delegationProducerService.ts

```diff
@@ -0,0 +1,289 @@
+import { delegationApi } from "pagopa-interop-api-clients";
+import {
+  AppContext,
+  DB,
+  eventRepository,
+  FileManager,
+  Logger,
+  PDFGenerator,
+  WithLogger,
+} from "pagopa-interop-commons";
+import {
+  CorrelationId,
+  Delegation,
+  DelegationId,
+  delegationEventToBinaryDataV2,
+  delegationKind,
+  EService,
+  delegationState,
+  EServiceId,
+  generateId,
+  Tenant,
+  TenantId,
+  unsafeBrandId,
+  WithMetadata,
+} from "pagopa-interop-models";
+import {
+  delegationNotFound,
+  eserviceNotFound,
+  tenantNotFound,
+} from "../model/domain/errors.js";
+import {
+  toCreateEventProducerDelegation,
+  toRevokeEventProducerDelegation,
+  toCreateEventApproveDelegation,
+  toCreateEventRejectDelegation,
+} from "../model/domain/toEvent.js";
+import { config } from "../config/config.js";
+import { ReadModelService } from "./readModelService.js";
+import {
+  assertDelegationIsRevokable,
+  assertDelegationNotExists,
+  assertDelegatorIsIPA,
+  assertDelegatorIsNotDelegate,
+  assertEserviceExists,
+  assertTenantAllowedToReceiveProducerDelegation,
+  assertIsDelegate,
+  assertIsState,
+} from "./validators.js";
+import { contractBuilder } from "./delegationContractBuilder.js";
+
+// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
+export function delegationProducerServiceBuilder(
+  dbInstance: DB,
+  readModelService: ReadModelService,
+  pdfGenerator: PDFGenerator,
+  fileManager: FileManager
+) {
+  const getTenantById = async (tenantId: TenantId): Promise<Tenant> => {
+    const tenant = await readModelService.getTenantById(tenantId);
+    if (!tenant) {
+      throw tenantNotFound(tenantId);
+    }
+    return tenant;
+  };
+
+  const retrieveDelegationById = async (
```

Comment:
I suggest to define this outside the function in order to be used from delegationService too